src/ast/function-object.ts

```diff
@@ -28,23 +29,22 @@ module.exports = class FunctionObject {
         }
       }
     });
-
     // Now we analyze the body with the local context. Note that recursion is
     // allowed, because we've already inserted the function itself into the
     // outer context, so recursive calls will be properly resolved during the
     // usual "outward moving" scope search. Of course, if you declare a local
     // variable with the same name as the function inside the function, you'll
     // shadow it, which would probably be not a good idea.
     if (this.body) {
-      this.body.forEach(s => s.analyze(context));
+      this.body.forEach((s: Statement) => s.analyze(context));
     }
   }
 
-  optimize() {
-    this.parameters.forEach(p => p.optimize());
-    this.body.forEach(s => s.optimize());
-    this.body = this.body.filter(s => s !== null);
+  public optimize(): FunctionObject {
+    this.params.forEach((p: Parameter) => p.optimize());
```

Comment:
You probably know TypeScript better than I do, but it would seem to me that because in your constructor, you said that the type of `this.params` was `Parameter[]`, that it would be proper for inference to kick in here and that you should leave the parameter of the anonymous function as `p` (without the type qualification).(Even Java infers there for its lambdas) Does that make sense? And ditto for all the other places like this.

Reply:
makes sense ! I think I needed to do it earlier while I was building out the types but now that the types are built out this this is great feedback for catching this out ! done !